### DISCUSSION TITLE
Instant fade skin option

### DISCUSSION THREAD

**Discussion by developomp** [ppy/osu#24018] [Ideas] [answered]
(this has nothing to do with #18789 or #5913)

## Who is this for?

In the osu!starndard ruleset, many top players such as [Karcher (peak #15)](https://osu.ppy.sh/users/11443437), [sytho (peak #33)](https://osu.ppy.sh/users/4175698), and of course [WubWoofWolf (peak #1)](https://osu.ppy.sh/users/39828) uses what's called an instant-fade skin where the hitcircles disappear without playing the fade animation.

## Why is this important

The way instant-face skin works is essentially a hack where what appears to be the hitcrcles are actually the number images, and the real hitcircle images are invisible (either it is 1x1 pixel in size or is transparent).

Because of this, many people are forced manually create a new skin just to have a instant-fade version of their favorite skin. People also rely on external tools such as [this one](https://github.com/clarks03/Osu-Instafade-Tool).

And because it is a hack, there are issues such as digits of high combo displaying on top of each other ([source](https://www.youtube.com/watch?v=mA38dTz6H5Q)).

![Screenshot from 2023-06-24 00-20-16](https://github.com/ppy/osu/assets/40858122/c077453d-67f6-402e-8970-3b98ce2f84bb)

The fade doesn't even appear to be instant either:

https://github.com/ppy/osu/assets/40858122/964fbf26-cf8c-4dcd-9a29-9e01a93a4de8

## What I propose

Add a config option in the skin for a proper instant fade hitcircle feature. If many people want a feature, they should not rely on a hack.

**Comment by monochrome22** [comment]
The situation at https://github.com/ppy/osu/issues/5913 is indeed a bit strange, but there was a follow up discussion where peppy gave an answer regarding adding insta fade: https://github.com/ppy/osu/discussions/20027.

**Reply by developomp** [reply]
Should probably change the feature request from a togglable setting in the game to a skin config option then.

**Reply by monochrome22** [reply]
A skin config option probably counts as "a hacky way".
I would imagine a proper way means adding it to the skin editor in some way.

**Comment by bdach** [comment]
I can't right now figure out where to find the discussions on this (because multiple were had) and the main thing here is that the instant fade stuff objectively looks terrible. I get that for the competitive 1% it may improve player performance, but that's not an excuse to kneecap visuals for the rest either.

You could pose questions as to whether allowing instafade at all isn't breaking game balance in some way, but given that stable supports this, the ship on that has sailed a long time ago, and we can't just remove support for it. However I kind of think that this setting shouldn't be very visibly exposed and should only be there for users that are really desperate to use it and the `hitcirclenumber` "hack" isn't the worst way to achieve that - and if support for this comes to lazer it should be at a similar level of hidden.

**Reply by developomp** [reply]
How about adding a undocumented option in the skin.ini file that makes the circles fade instantly?

**Reply by bdach** [reply]
Maybe. It won't be in `skin.ini` as legacy formats will not be extended with new functionality unless absolutely unavoidable.

**Reply by monochrome22** [reply]
>that's not an excuse to kneecap visuals for the rest either.

By "the rest" you mean spectators? 
I don't know if more worse visuals is a concern. There are already many insta fade skins that players can use and there are even tools that easily change an existing skin into insta fade. If someone doesn't already use an insta fade skin, then I feel it's unlikely they would use a setting for it. If someone is competitive and trying to squeeze out as much performance as possible, then they would already be using an insta fade skin, like how many people use Rafis skin for DT.
This argument is weird in general. Basically, I shouldn't be allowed to play the game the way that I want to because someone else might think that my gameplay doesn't look good? Very strange. It's not just about improving performance; insta fade feels better for me. I would likely use it even if it didn't make reading more comfortable. Though I agree visually it can look worse, particularly on streams. Wouldn't say "objectively terrible", but at times not as good as regular fading, sure.

If worse visuals really are a concern, then as an insta fade user, personally I'm fine with it being supported in a hidden way. I just want to have proper combo numbers and colors on my skin and not have to change skins when I'm playing with custom beatmap skins, since insta fade will cause the hitcircles to be invisible on custom skins. Though a slider from 0 to 480 for fade_duration would be better. If other players not wanting to see bad visuals in-game is a concern, then this slider should be in settings instead of the skin editor. And you shouldn't be concerned with what someone's gameplay looks like outside of the game (YouTube videos, livestreams etc).
By the way, the only way a non-insta fade user would see insta fade in-game would be through downloading the skin while spectating, as said in https://github.com/ppy/osu/discussions/21729#discussioncomment-4455974. So, by adding a slider in settings, you would actually be reducing the amount of people who see insta fade, since they would have the option to turn it off and not be forced to spectate with insta fade on with it being built into the skin.
Though I don't know [what solution peppy had in mind](https://github.com/ppy/osu/discussions/20027#discussioncomment-3507125), maybe there's a much better way than a slider.

**Reply by bdach** [reply]
> By "the rest" you mean spectators?

Spectators, beginner players, etc.

> I don't know if more worse visuals is a concern. There are already many insta fade skins that players can use and there are even tools that easily change an existing skin into insta fade.

This works both ways - if tooling around this sort of thing has already developed, then does it need to be supported in the game itself?

> This argument is weird in general. Basically, I shouldn't be allowed to play the game the way that I want to because someone else might think that my gameplay doesn't look good?

That's misconstruing what I'm saying. You *are* allowed to ~~make your skin uglier~~ play the game you want to, it's just that it's considered an "expert" option if you will and as such not exposed to the general player base. Not everyone is an ultra competitive player.

> Though a slider from 0 to 480 for fade_duration would be better.

Some people want everything to be a slider and we cannot work with that. Some limits to customisability must be in place. Else it expands the support burden because the game must work correctly for any old combination of sliders.

---

I see this feature of an analogue of source engine console variables like `lod_bias` or `viewmodel_fov` or whatever. You can use them and make your game ~~look terrible~~ have less visual noise, but most of those settings aren't exposed to the general player base, *and that's completely fine*.

**Reply by monochrome22** [reply]
> That's misconstruing what I'm saying.

I understand what you're saying, I was just pointing out how "it's not an excuse to kneecap visuals for the rest" is a really strange argument. 

> This works both ways - if tooling around this sort of thing has already developed, then does it need to be supported in the game itself?

Yes, because even with the tooling there are issues (which have been pointed out). You would want the new and good osu version to fix the issues around your preferred playstyle and provide a better experience.

> it's just that it's considered an "expert" option if you will and as such not exposed to the general player base.

Honestly I'm probably too biased to comment on this. I've played with insta fade for years, so to me it seems like a common mechanic and not a rare expert option that only ultra competitive players use. 
Even so, I will say that I understand there can't be an option for every little thing, but insta fade seems more important.
Will see what other people think.

In any case, I'm personally fine with any implementation that fixes the issues.

**Reply by ghost** [reply]
This has been a wanted feature for years and despite the "ship already being sailed" on these skins, you are ignoring it because you just dont want to do it? Of course people are going to want to disable a special effect in a competitive game. Why would you intentionally mess up the player experience just because you personally think it's ugly to not have a special effect? Many people play with pro skins and instafade ones are some of the most popular for a reason. Beginner users find them and use them also because they are at the top of all of the skins websites they are directed to. Why is there so much resistance to allowing proper combo numbers and colors like every other skin? It doesn't make any sense to me.

**Reply by egucchi** [reply]
You thinking it looks terrible really doesn't matter, nobody is gonna force you to enable the option dude. Let people have their shit how they want it, why does your opinion on the way something looks (that remember, that you won't even be using) matter at all when it's an OPTIONAL thing for the people who want to use it?

**Reply by monochrome22** [reply]
Just to be clear: nobody is against adding insta fade. The discussion is only about _how_ it should be added. bdach is saying it should be a "hidden" option, something similar to a line in skin.ini, [peppy](https://github.com/ppy/osu/discussions/20027#discussioncomment-3507125) is saying it should be added in "not a hacky way", don't know what he has in mind, and I'm saying it should be fine to have it as a slider in options.
I guess the way forward would be determined by how many players actually use insta fade. If it's used by many, then it should be a slider in settings, if it's not really that popular, then it would be fine to have it be more hidden. 

>despite the "ship already being sailed" on these skins

When he said "the ship has sailed", he meant there's no point having a discussion about whether or not insta fade should be disabled due to balance concerns, since there's already too many scores that used insta fade.

**Reply by nitedani** [reply]
"Instafade" objectively makes reading easier, removing visual clutter and mental overhead. I think it is especially important to minimize special effects in a game like osu.

**Reply by mouceen** [reply]
> "Instafade" objectively makes reading easier, removing visual clutter and mental overhead. I think it is especially important to minimize special effects in a game like osu.

Not true. Would you say the same for hidden? It's preference, hidden even gives more PP, and removes visual clutter at the same time.

> and the main thing here is that the instant fade stuff objectively looks terrible.

Visuals are mostly subjective, unless something is really really crap, its not objectively terrible. I think theres nothing wrong visually with instafading circles. A small button to enable it in settings would be amazing. There could be a greater argument that hit lighting should be disabled automatically and moved as a hidden feature because nobody who got past 2 star diffs uses it.

**Comment by Toasterly** [comment]
PR attempting to add this feature as nothing has changed in years:
https://github.com/ppy/osu/pull/33546

**Comment by developomp** [comment, marked as answer]
https://github.com/ppy/osu/pull/38371
